doc/neps/nep-0055-string_dtype.rst

```diff
@@ -656,13 +518,385 @@ string validation in a separate, expensive, pass over the input array-like. We
+*************************
+
+Each entry in a ``StringDType`` array will be defined to be an instance of an
+opaque ``npy_packed_static_string`` struct. Making the public definition opaque
+will allow us to in principle change the underlying C implementation or even the
+size of the struct in the future.
+
+To actually work with the string data, we will publicly expose a typedef for
+a struct with with the following definition:
+
+.. code-block:: C
+
+   struct npy_static_string {
+       size_t size;
+       const char *buf;
+   };
+
+Where ``size`` is the size, in bytes, of the string and ``buf`` is a const
+pointer to the beginning of a UTF-8 encoded bytestream containing string
+data. This is a *read-only* view onto the string, we will not expose a public
+interface for modifying these strings. We do not append a trailing null
+character to the byte stream, so users attempting to pass the ``buf`` field to
+an API expecting a C string must create a copy with a trailing null. As a
+positive consequence, ``StringDType`` array entries can contain arbitrary
+embedded or trailing null characters.
+
+Say we are writing a ufunc implementation for ``StringDType``. If we are given
+``const char *buf`` pointer to the beginning of a ``StringDType`` array entry, and a
+``PyArray_Descr *`` pointer to the array descriptor, one can
+access the underlying string data like so:
+
+.. code-block:: C
+
+   npy_string_allocator *allocator = NpyString_acquire_allocator(descr);
+
+   npy_static_string sdata = {0, NULL};
+   npy_packed_static_string *packed_string = (npy_packed_static_string *)buf;
+   int is_null = 0;
+
+   if (NpyString_load(allocator, packed_string, &sdata) == -1) {
+       // failed to load string, set error
+       return -1;
+   }
+
+   if (is_null) {
+       // handle missing string
+       // sdata->buf is NULL
+       // sdata->size is 0
+   }
+   else {
+       // sdata->buf is a pointer to the beginning of a string
+       // sdata->size is the size of the string
+   }
+   NpyString_release_allocator(descr);
+
+We will also expose variants of these functions that take two or three
+descriptors for situations where several different allocators need to be
+simultaneously acquired. These variants simultaneously handle the case where the
+input descriptors are the same object or different objects. In the first case,
+only one allocator needs to be acquired, but in the latter case multiple
+allocators need to be acquired. The user will only need to think about choosing
+the function to acquire or release allocators based on the number of descriptors
+they need to work with.
+
+Similarly, we will also expose ``NpyString_pack`` to create a new
+packed string in an existing array from the contents of a ``char *`` buffer:
+
+.. code-block:: C
+
+   char *str = "Hello world";
+   size_t size = 11;
+   npy_packed_static_string *packed_string = (npy_packed_static_string *)buf;
+
+   npy_string_allocator *allocator = NpyString_acquire_allocator(descr);
```

Comment:
Will the allocator be opaque for now (I guess you don't say).

Reply:
yup, it's opaque, that should be clearer now